cpp/src/gandiva/precompiled/string_ops.cc

```diff
@@ -1642,6 +1642,55 @@ const char* convert_toUTF8(int64_t context, const char* value, int32_t value_len
   return value;
 }
 
+// Calculate the levenshtein distance between two string values
+FORCE_INLINE
+gdv_int32 levenshtein_utf8_utf8(int64_t context, const char* in1, int32_t in1_len,
+                                const char* in2, int32_t in2_len) {
+  if (in1_len < 0 || in2_len < 0) {
+    gdv_fn_context_set_error_msg(context, "String length must be greater than 0");
+    return 0;
+  }
+  int32_t a, b, c, aux;
+  int32_t len_dist_1 = in1_len + 1;
+  int32_t len_dist_2 = in2_len + 1;
+  // dist[i][j] represents the Levenstein distance between the strings
+  int** dist;
+  if ((dist = (int**)malloc((len_dist_1 + 1) * sizeof(int*))) == NULL) {
+    gdv_fn_context_set_error_msg(context, "Insufficient space to allocate buffer");
+    return 0;
+  }
+  for (int i = 0; i <= in1_len; i++) {
+    if ((dist[i] = (int*)malloc(len_dist_2 * sizeof(int*))) == NULL) {
```

Comment:
Can we allocate one large chunk of memory instead of making multiple calls to malloc? Each call to malloc requires a lock!!!

* ptr = malloc((in1_len + 1) * (in2_len + 1) * sizeof(int)) space. Line 1663 is allocating space for int*
* rowSize = (int2_len + 1) * sizeof(int)
* dist[i] = ptr + i * rowSize

Reply:
Oh, i don't have did think in this, i'll trying modify

Reply:
@ViniciusSouzaRoque I have reviewed levenshtien in the following pull request:

https://github.com/apache/arrow/pull/11051

Why there are two pull requests? Can you remove one of them? Also, my comments are similar to Vivek, but there might be additional comments there and a link to a memory efficient levenstein implementation. See if they are helpful.